src/plugins/walletConnectToDapps/WalletConnectBridgeProvider.tsx

```diff
@@ -0,0 +1,500 @@
+      const signedMessage = await chainAdapter.signMessage(input)
+      if (!signedMessage) throw new Error('WalletConnectBridgeProvider: signMessage failed')
+      return signedMessage
+    },
+    [accountMetadataById, wallet, wcAccountId],
+  )
+
+  const signTypedData = useCallback(
+    async (hashableData: TypedData) => {
+      if (!hashableData) return
+      if (!wallet) return
+      if (!wcAccountId) return
+      const accountMetadata = accountMetadataById[wcAccountId]
+      if (!accountMetadata) return
+      const { bip44Params } = accountMetadata
+      const addressNList = toAddressNList(bip44Params)
+      const messageToSign = { addressNList, hashableData }
+      if (wallet instanceof KeepKeyHDWallet || wallet instanceof NativeHDWallet) {
+        const signedMessage = await wallet?.ethSignTypedData(messageToSign)
+        if (!signedMessage) throw new Error('WalletConnectBridgeProvider: signTypedData failed')
+        console.log({ signedMessage })
+        debugger
+        return signedMessage.signature
+      }
+    },
+    [accountMetadataById, wallet, wcAccountId],
+  )
+
+  const handleSessionRequest = useCallback(
+    (error: Error | null, payload: WalletConnectSessionRequestPayload) => {
+      if (error) moduleLogger.error(error, { fn: '_onSessionRequest' }, 'Error session request')
+      if (!connector) return
+      if (!wcAccountId) return
+      const { chainId, account } = fromAccountId(wcAccountId)
+      moduleLogger.info(payload, 'approve wc session')
+      connector.approveSession({
+        chainId: parseInt(fromChainId(chainId).chainReference),
+        accounts: [account],
+      })
+    },
+    [connector, wcAccountId],
+  )
+
+  const approveRequest = useCallback(
+    async (
+      request: WalletConnectCallRequest,
+      approveData: Partial<
+        WalletConnectCallRequestResponseMap[keyof WalletConnectCallRequestResponseMap]
+      >,
+    ) => {
+      if (!wallet) return
+      if (!wcAccountId) return
+      if (!connector) return
+
+      const maybeChainAdapter = getChainAdapterManager().get(fromAccountId(wcAccountId).chainId)
+      if (!maybeChainAdapter) return
+      const chainAdapter = maybeChainAdapter as unknown as EvmBaseAdapter<EvmChainId>
+
+      // TODO(0xdef1cafe): IIFE
+      let result: any
+      switch (request.method) {
+        case 'eth_sign': {
+          result = await signMessage(convertHexToUtf8(request.params[1]))
+          break
+        }
+        case 'eth_signTypedData': {
+          const payloadString = request.params[1]
+          const typedData = JSON.parse(payloadString)
+          const signed = await signTypedData(typedData)
+          if (!signed) break
+          result = signed
+          break
+        }
+        case 'personal_sign': {
+          result = await signMessage(convertHexToUtf8(request.params[0]))
+          break
+        }
+        case 'eth_sendTransaction': {
+          const advancedParams = approveData as ETHSignTx
```

Comment:
this function is a monster. we should pull the biz logic out of each of these cases into smaller handlers

Reply:
Precisely `this`